servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,257 @@
+                         final BufferAllocator allocator) {
+        requireNonNull(src);
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0)");
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderFn.get();
+        final EmbeddedChannel channel = new EmbeddedChannel(encoder);
+
+        try {
+            ByteBuf origin = toByteBufRetained(src, offset, length);
+            channel.writeOutbound(origin);
+
+            // May produce footer
+            channel.close();
+
+            CompositeBuffer compositeBuffer = allocator.newCompositeBuffer();
+            ByteBuf chunk;
+            while ((chunk = channel.readOutbound()) != null) {
+                compositeBuffer.addBuffer(newBufferFromRetained(chunk));
+            }
+
+            return compositeBuffer;
+        } catch (Exception e) {
+            LOGGER.error("Error while encoding with {}", name(), e);
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeClose(channel);
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> encode(final Publisher<Buffer> from,
+                                          final BufferAllocator __) {
+        requireNonNull(from);
+        return from
+                .concat(succeeded(END_OF_STREAM))
+                .liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+                    private final MessageToByteEncoder<ByteBuf> encoder = encoderFn.get();
+                    private final EmbeddedChannel channel = new EmbeddedChannel(encoder);
+
+                    @Override
+                    public void onSubscribe(PublisherSource.Subscription subscription) {
+                        subscriber.onSubscribe(subscription);
+                    }
+
+                    @Override
+                    public void onNext(Buffer next) {
+                        if (next == null) {
+                            return;
+                        }
+
+                        // onNext will produce AT-MOST N items (from upstream)
+                        // +1 for the encoding footer (ie. END_OF_STREAM)
+                        if (next == END_OF_STREAM) {
+                            // May produce footer
+                            channel.close();
+                            ByteBuf part = channel.readOutbound();
+                            if (part != null) {
+                                subscriber.onNext(newBufferFromRetained(part));
+                            }
+
+                            return;
+                        }
+
+                        channel.writeOutbound(toByteBufRetained(next));
+                        ByteBuf part = channel.readOutbound();
+                        if (part != null) {
```

Comment:
Here is a risk that some codecs may produce more than one outbound `ByteBuf`. We should draint it until the outbound messages queue is empty here and for `END_OF_STREAM`.